### PR TITLE
Allow systems with arrays within variables

### PR DESCRIPTION
**Pull request by richfitz**
Fairly small in the end, now that we have the packing stuff sorted from mcstate2.  This PR allows a system to have arrays within variables.  The next one will allow vector parameters, then after that variable sized arrays

### CODE REVIEW

File: R/util.R [hunk 1]
```diff
@@ -115,10 +115,22 @@ uses_unary_minus <- function(expr) {
...
     a - b
   } else {
     call("-", a, b)
   }
 }
+
+
+expr_plus <- function(a, b) {
+  if (is.numeric(a) && is.numeric(b)) {
+    a + b
```
Comment (weshinsley) [inline R/util.R:132]: Code coverage - this numeric add doesn't get covered - will try and suggest something below...

File: R/generate_dust.R [hunk 2]
```diff
@@ -448,20 +438,23 @@ generate_dust_lhs <- function(lhs, dat, name_state, options) {
...
     sprintf("%s[%s]",
-            name_state, match(name, dat$storage$packing$state$scalar) - 1)
+            name_state,
+            generate_dust_sexp(offset, dat$sexp_data, options))
   } else if (location == "adjoint") {
+    packing <- dat$storage$packing$adjoint
+    offset <- packing$offset[[match(name, packing$name)]]
     if (is_array) {
-      stop("support arrays in adjoint")
+      offset <- expr_plus(idx, offset)
```
Comment (weshinsley) [inline R/generate_dust.R:453]: Not covered

File: R/parse_system.R [hunk 3]
```diff
@@ -360,3 +357,30 @@ parse_system_arrays <- function(exprs, call) {
...
+
+parse_packing <- function(names, arrays) {
+  scalar <- setdiff(names, arrays$name)
+  if (length(scalar) > 0) {
+    packing_scalar <- data_frame(
+      name = scalar, rank = 0, dims = I(vector("list", length(scalar))),
+      size = I(rep(list(1), length(scalar))))
+    packing <- rbind(packing_scalar, arrays)
+  } else {
+    packing <- arrays
```
Comment (weshinsley) [inline R/parse_system.R:370]: Not covered yet

**Review by weshinsley — COMMENTED** [review]
Slight rabbit-hole maybe - but missing code coverage on a couple of lines in this PR that might not be important - 

generate_dust.R:453 - where location - "adjoint" and is_array is true - BUT, also missing coverage on the numeric part of expr_plus, so I wonder if this is worth a look

parse_system.R:379 - the null case probably where length(scalar) == 0

util.R:132 - numeric expr_plus

Not in this PR, and probably not important, but...

parse_compat.R:98 - where parse_add_compat is called with expr$compat already set, and appends the new info. (I am not sure how this can be triggered?)

File: R/generate_dust.R (hunk 2 above)
Comment (weshinsley) [reply R/generate_dust.R:453]:
```
test_that("expr_plus", {
  expect_identical(expr_plus(1, 2), 3)
  x <- expr_plus(quote(var), 2)
  expect_type(x, "language")
  x <- as.character(x)
  expect_equal(x, c("+", "var", "2"))
})
```

**Comment by richfitz** [comment]
> parse_compat.R:98 - where parse_add_compat is called with expr$compat already set, and appends the new info. (I am not sure how this can be triggered?)

The next PR (#27) has a good trigger for this, I'll turn it into a test case

File: R/generate_dust.R (hunk 2 above)
Comment (richfitz) [reply R/generate_dust.R:454]: this one here I think needs to wait as we prevent use of differentiation in systems with arrays - I've put an assertion in now above the uncovered line to mark that

File: R/parse_system.R (hunk 3 above)
Comment (richfitz) [reply R/parse_system.R:370]: covered now

**Review by weshinsley — APPROVED** [review]

**Pull request merged**